File: SmartLink/Source/SmartLinkClient/NRC_ServiceGenericClasses/TaskSchedulerClasses/TaskScheduler_HourlyScheduler.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
//********************************************************************'
// Created by Elibad - 09/01/2007
//
//  This class inheriths the TaskScheduler functionality
//  and implements the Hourly Schedule
//*********************************************************************
namespace NRC.Miscellaneous.TaskScheduler
{
	public class HourlyScheduler : Scheduler
	{

		public HourlyScheduler() : base(1, System.DateTime.Now.ToString(), System.DateTime.Now.ToString(), false)
		{
		}

		public HourlyScheduler(int Interval, string StartDate, string Time, bool Enabled) : base(Interval, StartDate, Time, Enabled)
		{
		}

		protected override System.DateTime GetNextOccurrence(System.DateTime BaseDate)
		{
			System.DateTime functionReturnValue = default(System.DateTime);
			if (this._bFirstRun) {
				System.DateTime dResult = System.DateTime.Parse(BaseDate.ToString("MM/dd/yyyy") + " " + this.Time);
				if (this.Time.CompareTo(System.DateTime.Now.ToString("HH:mm:ss")) < 0) {
					if (dResult.ToString("mm").CompareTo(System.DateTime.Now.ToString("mm")) <= 0) {
						dResult = DateTime.Parse(dResult.ToString("MM/dd/yyyy ") + System.DateTime.Now.AddHours(1).ToString("HH:") + dResult.ToString("mm:ss"));
					} else {
						dResult = DateTime.Parse(dResult.ToString("MM/dd/yyyy ") + System.DateTime.Now.ToString("HH:") + dResult.ToString("mm:ss"));
					}
				}
				functionReturnValue = dResult;
			} else {
				functionReturnValue = BaseDate.AddHours(this.Interval);
			}
			this._bFirstRun = false;
			return functionReturnValue;
		}

	}

}

File: SmartLink/Source/SmartLinkClient/NRC_ServiceGenericClasses/TaskSchedulerClasses/TaskScheduler_HourlyScheduler.vb
'********************************************************************'
' Created by Elibad - 09/01/2007
'
'  This class inheriths the TaskScheduler functionality
'  and implements the Hourly Schedule
'*********************************************************************
Namespace Miscellaneous.TaskScheduler
    Public Class HourlyScheduler : Inherits Scheduler

        Public Sub New()
            MyBase.New(1, Date.Now.ToString, Date.Now.ToString, False)
        End Sub

        Public Sub New(ByVal Interval As Integer, ByVal StartDate As String, ByVal Time As String, ByVal Enabled As Boolean)
            MyBase.New(Interval, StartDate, Time, Enabled)
        End Sub

        Protected Overrides Function GetNextOccurrence(ByVal BaseDate As Date) As Date
            If Me._bFirstRun Then
                Dim dResult As Date = Date.Parse(BaseDate.ToString("MM/dd/yyyy") & " " & Me.Time)
                If Me.Time < Date.Now.ToString("HH:mm:ss") Then
                    If dResult.ToString("mm") <= Date.Now.ToString("mm") Then
                        dResult = DateTime.Parse(dResult.ToString("MM/dd/yyyy ") & Date.Now.AddHours(1).ToString("HH:") & dResult.ToString("mm:ss"))
                    Else
                        dResult = DateTime.Parse(dResult.ToString("MM/dd/yyyy ") & Date.Now.ToString("HH:") & dResult.ToString("mm:ss"))
                    End If
                End If
                GetNextOccurrence = dResult
            Else
                GetNextOccurrence = BaseDate.AddHours(Me.Interval)
            End If
            Me._bFirstRun = False
        End Function

    End Class

End Namespace
